Userland/Libraries/LibWeb/Streams/AbstractOperations.cpp

```diff
@@ -1068,13 +1363,135 @@ WebIDL::ExceptionOr<void> writable_stream_default_controller_error(WritableStrea
+
+    // 2. Perform ! WritableStreamMarkFirstWriteRequestInFlight(stream).
+    writable_stream_mark_first_write_request_in_flight(*stream);
+
+    // 3. Let sinkWritePromise be the result of performing controller.[[writeAlgorithm]], passing in chunk.
+    auto sink_write_promise = TRY((*controller.write_algorithm())(chunk));
+
+    // 4. Upon fulfillment of sinkWritePromise,
+    WebIDL::upon_fulfillment(*sink_write_promise, [&, stream = stream](auto const&) -> WebIDL::ExceptionOr<JS::Value> {
+        // 1. Perform ! WritableStreamFinishInFlightWrite(stream).
+        writable_stream_finish_in_flight_write(*stream);
+
+        // 2. Let state be stream.[[state]].
+        auto state = stream->state();
+
+        // 3. Assert: state is "writable" or "erroring".
+        VERIFY(state == WritableStream::State::Writable || state == WritableStream::State::Erroring);
+
+        // 4. Perform ! DequeueValue(controller).
+        dequeue_value(controller);
+
+        // 5. If ! WritableStreamCloseQueuedOrInFlight(stream) is false and state is "writable",
+        if (!writable_stream_close_queued_or_in_flight(*stream) && state == WritableStream::State::Writable) {
+            // 1. Let backpressure be ! WritableStreamDefaultControllerGetBackpressure(controller).
+            auto backpressure = writable_stream_default_controller_get_backpressure(controller);
+
+            // 2. Perform ! WritableStreamUpdateBackpressure(stream, backpressure).
+            writable_stream_update_backpressure(*stream, backpressure);
+        }
+
+        // 6 .Perform ! WritableStreamDefaultControllerAdvanceQueueIfNeeded(controller).
+        TRY(writable_stream_default_controller_advance_queue_if_needed(controller));
+
+        return JS::js_undefined();
+    });
+
+    // 5. Upon rejection of sinkWritePromise with reason,
+    WebIDL::upon_rejection(*sink_write_promise, [&, stream = stream](auto const& reason) -> WebIDL::ExceptionOr<JS::Value> {
+        // 1. If stream.[[state]] is "writable", perform ! WritableStreamDefaultControllerClearAlgorithms(controller).
+        if (stream->state() == WritableStream::State::Writable)
+            writable_stream_default_controller_clear_algorithms(controller);
+
+        // 2. Perform ! WritableStreamFinishInFlightWriteWithError(stream, reason).
+        TRY(writable_stream_finish_in_flight_write_with_error(*stream, reason));
+
+        return JS::js_undefined();
+    });
+
+    return {};
+}
+
+// https://streams.spec.whatwg.org/#close-sentinel
+// Non-standard function that implements the "close sentinel" value.
+JS::Value create_close_sentinel()
+{
+    // The close sentinel is a unique value enqueued into [[queue]], in lieu of a chunk, to signal that the stream is closed. It is only used internally, and is never exposed to web developers.
+    // Note: We use the empty Value to signal this as, similarly to the note above, the empty value is not exposed to nor creatable by web developers.
+    return {};
```

Comment:
@linusg I know you mentioned you didn't like using the empty value here, do you have any other ideas? The only thing I can think of it using an `Object*` with a special address like `0x1` or something, not sure if that'd be a better choice.

Reply:
It's fine, I just dislike the empty value concept, but it's not like we don't already use it in a bunch of places.